⚠️ Bump CAPI to v1.5.0

Sunnatillo opened the pull request:
🌱 This PR uplifts:
  1. CAPI to v1.5.0
  2. CR to v0.15.1
  3. Controller-tools to v0.12.0

Also: 

- Uplifts IPAM to main branch (temporarily, until next release of IPAM)
- Adds `WithStatusSubresource` to fake client, due to status handling changes in [CR v0.15.0](https://github.com/kubernetes-sigs/controller-runtime/pull/2259)
- Adds `finalizer`s to objects with `deletionTimeStamp` due to changes in [CR v0.15.0](https://github.com/kubernetes-sigs/controller-runtime/pull/2316)
- Adjusts webhooks to support warnings due to change in [CR v0.15.0](https://github.com/kubernetes-sigs/controller-runtime/pull/2014)
- Introduces CAPI_CONFIG_FOLDER variable since [CAPI started following XDG standard](https://github.com/kubernetes-sigs/cluster-api/pull/6913)

Sunnatillo commented:
/test-centos-e2e-integration-main
/test-ubuntu-e2e-integration-main
/test-ubuntu-integration-main
/test-centos-integration-main
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-integration-main
/test-ubuntu-integration-main

Sunnatillo commented:
/test-ubuntu-integration-main

Sunnatillo commented:
/test-centos-e2e-integration-main
/test-ubuntu-integration-main

Sunnatillo commented:
/test-ubuntu-e2e-integration-main
/test-centos-integration-main
/test-centos-e2e-feature-main
/test-e2e-upgrade-main-from-release-1-4
/test-e2e-1-26-1-27-upgrade-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/keep-test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-integration-main
/test-ubuntu-e2e-integration-main
/test-ubuntu-integration-main
/test-centos-integration-main
/test-e2e-upgrade-main-from-release-1-4
/test-e2e-1-26-1-27-upgrade-main

kashifest commented on baremetal/metal3datatemplate_manager_test.go:
```diff
@@ -352,6 +352,7 @@ var _ = Describe("Metal3DataTemplate manager", func() {
 						Name:              "abcf",
 						Namespace:         namespaceName,
 						DeletionTimestamp: &timeNow,
+						Finalizers:        []string{"ipclaim.ipam.metal3.io"},
```
trying to understand this addition, if its was missing how were we testing the finalizer previously.

Sunnatillo replied:
This change is due to [CR 0.15 related](https://github.com/kubernetes-sigs/controller-runtime/pull/2316). Now CR does not accept objects with `deletionTImeStamp` without `Finalizers`

kashifest commented on config/crd/bases/infrastructure.cluster.x-k8s.io_metal3clusters.yaml:
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.11.4
+    controller-gen.kubebuilder.io/version: v0.12.1
```
this as actually one patch version ahead of CAPI 1.5 branch but I guess this is fine.

kashifest commented on controllers/metal3data_controller.go:
```diff
@@ -160,7 +159,7 @@ func (r *Metal3DataReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Ma
 // Metal3IPClaimToMetal3Data will return a reconcile request for a Metal3Data if the event is for a
 // Metal3IPClaim and that Metal3IPClaim references a Metal3Data.
-func (r *Metal3DataReconciler) Metal3IPClaimToMetal3Data(obj client.Object) []ctrl.Request {
+func (r *Metal3DataReconciler) Metal3IPClaimToMetal3Data(_ context.Context, obj client.Object) []ctrl.Request {
```
we dont need the ctx var anymore ? why are we not defining it?

kashifest commented on controllers/metal3datatemplate_controller.go:
```diff
@@ -181,7 +180,7 @@ func (r *Metal3DataTemplateReconciler) SetupWithManager(ctx context.Context, mgr
 // Metal3DataTemplate if the event is for a
 // Metal3DataClaim and that Metal3DataClaim references a Metal3DataTemplate.
-func (r *Metal3DataTemplateReconciler) Metal3DataClaimToMetal3DataTemplate(obj client.Object) []ctrl.Request {
+func (r *Metal3DataTemplateReconciler) Metal3DataClaimToMetal3DataTemplate(_ context.Context, obj client.Object) []ctrl.Request {
```
same

kashifest commented on controllers/metal3machine_controller.go:
```diff
@@ -423,7 +422,7 @@ func (r *Metal3MachineReconciler) Metal3ClusterToMetal3Machines(o client.Object)
 // BareMetalHostToMetal3Machines will return a reconcile request for a Metal3Machine if the event is for a
 // BareMetalHost and that BareMetalHost references a Metal3Machine.
-func (r *Metal3MachineReconciler) BareMetalHostToMetal3Machines(obj client.Object) []ctrl.Request {
+func (r *Metal3MachineReconciler) BareMetalHostToMetal3Machines(_ context.Context, obj client.Object) []ctrl.Request {
```
same

kashifest commented on controllers/metal3machine_controller.go:
```diff
@@ -480,7 +479,7 @@ func (r *Metal3MachineReconciler) Metal3DataClaimToMetal3Machines(obj client.Obj
 // Metal3DataToMetal3Machines will return a reconcile request for a Metal3Machine if the event is for a
 // Metal3Data and that Metal3Data references a Metal3Machine.
-func (r *Metal3MachineReconciler) Metal3DataToMetal3Machines(obj client.Object) []ctrl.Request {
+func (r *Metal3MachineReconciler) Metal3DataToMetal3Machines(_ context.Context, obj client.Object) []ctrl.Request {
```
same

kashifest commented on controllers/metal3machinetemplate_controller.go:
```diff
@@ -130,7 +129,7 @@ func (r *Metal3MachineTemplateReconciler) SetupWithManager(ctx context.Context,
 // Metal3MachinesToMetal3MachineTemplate is a handler.ToRequestsFunc to be used to enqeue
 // requests for reconciliation of Metal3MachineTemplates.
-func (r *Metal3MachineTemplateReconciler) Metal3MachinesToMetal3MachineTemplate(o client.Object) []ctrl.Request {
+func (r *Metal3MachineTemplateReconciler) Metal3MachinesToMetal3MachineTemplate(_ context.Context, o client.Object) []ctrl.Request {
```
same

kashifest commented on go.mod:
```diff
@@ -7,35 +7,35 @@ require (
+	k8s.io/klog/v2 v2.90.1
+	k8s.io/utils v0.0.0-20230209194617-a36077c30491
+	sigs.k8s.io/cluster-api v1.5.0
+	sigs.k8s.io/controller-runtime v0.15.0
```
check previous comment on the same module

kashifest commented on scripts/ci-e2e.sh:
```diff
@@ -8,14 +8,29 @@ export CAPM3PATH="${REPO_ROOT}"
+    CONFIG_FOLDER="${XDG_CONFIG_HOME:-$HOME/.config}"
+    export CAPI_CONFIG_FOLDER="${CONFIG_FOLDER}/cluster-api" 
+else
+    export CAPI_CONFIG_FOLDER="${HOME}/.cluster-api"
```
we are setting the same var twice in two consecutive lines?

Sunnatillo replied:
In the first line we are getting `XDG_CONFIG_HOME` variable as [CONFIG_FOLDER](url), if it is not set we are defaulting to  `$HOME/.config folder`. 
In the second line we are getting `CAPI_CONFIG_FOLDER` not `CONFIG_FOLDER`. 
For more please check https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
and https://github.com/kubernetes-sigs/cluster-api/pull/6913

kashifest replied:
I somehow didnt see the else :D , sorry ignore this comment.

kashifest reviewed (changes requested):
In addition to the comments in line, you have to add contract in these two files also 
1. https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/metadata.yaml
2. https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/test/e2e/data/infrastructure-metal3/main/metadata.yaml

Sunnatillo commented:
/test-centos-e2e-integration-main
/test-ubuntu-e2e-integration-main
/test-ubuntu-integration-main
/test-centos-integration-main
/test-centos-e2e-feature-main
/test-e2e-upgrade-main-from-release-1-4
/test-e2e-1-26-1-27-upgrade-main

Sunnatillo commented:
/test-centos-e2e-integration-main
/test-ubuntu-e2e-integration-main
/test-ubuntu-integration-main
/test-centos-integration-main
/test-centos-e2e-feature-main
/test-e2e-upgrade-main-from-release-1-4
/test-e2e-1-26-1-27-upgrade-main

Sunnatillo commented:
/test-centos-e2e-integration-main
/test-ubuntu-e2e-integration-main
/test-ubuntu-integration-main
/test-centos-integration-main
/test-centos-e2e-feature-main
/test-e2e-upgrade-main-from-release-1-4
/test-e2e-1-26-1-27-upgrade-main

Sunnatillo commented:
/test-centos-e2e-integration-main
/test-ubuntu-e2e-integration-main

kashifest commented:
lgtm
I will approve it once the tests pass.

Sunnatillo commented:
/keep-test-centos-e2e-integration-main

Sunnatillo commented:
/keep-test-ubuntu-e2e-integration-main

Sunnatillo commented:
/test-ubuntu-e2e-integration-main

Sunnatillo commented:
/test-ubuntu-e2e-integration-main
/test-centos-e2e-integration-main

Sunnatillo commented:
/test-ubuntu-e2e-integration-main
/test-centos-e2e-integration-main

Sunnatillo commented:
> In addition to the comments in line, you have to add contract in these two files also
> 
> 1. https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/metadata.yaml
> 2. https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/test/e2e/data/infrastructure-metal3/main/metadata.yaml

Why we need to add 1.5 to metadata when we do not have that release? It was failing when I added, and it passed failing point when I removed these 1.5 releaseSeries from metadata.

Sunnatillo commented:
/test-centos-e2e-integration-main
/test-ubuntu-integration-main
/test-centos-integration-main
/test-e2e-upgrade-main-from-release-1-4
/test-e2e-1-26-1-27-upgrade-main

Sunnatillo commented:
/test-ubuntu-integration-main
/test-centos-integration-main
/test-e2e-upgrade-main-from-release-1-4
/test-e2e-1-26-1-27-upgrade-main

furkatgofurov7 commented on api/go.mod:
```diff
@@ -4,30 +4,31 @@ go 1.20
+	k8s.io/apimachinery v0.27.2
+	k8s.io/client-go v0.27.2
+	k8s.io/utils v0.0.0-20230209194617-a36077c30491
 	sigs.k8s.io/cluster-api v1.4.2
```
why is this still `v1.4.2`?

Sunnatillo replied:
fixed

furkatgofurov7 commented on api/go.mod:
```diff
@@ -4,30 +4,31 @@ go 1.20
 require (
 	github.com/google/gofuzz v1.2.0
-	github.com/metal3-io/ip-address-manager/api v1.4.1
+	github.com/metal3-io/ip-address-manager/api v1.4.1-0.20230801074326-623678ce1ae7
```
I would suggest bumping IPAM to CAPI v1.5.0 first, release IPAM, come back here and bump IPAM to that new version for easier dependency cycle

Sunnatillo replied:
We already uplifted IPAM and specifying uplifted commit here, since there was not a release on ipam yet

furkatgofurov7 commented:
/retitle ⚠️ Bump CAPI to v1.5.0

omitting CR bump in the PR title due to it is a dependency of CAPI and comes with it automatically

Sunnatillo commented:
/test-centos-e2e-integration-main

Sunnatillo commented:
/test-centos-e2e-integration-main

Sunnatillo commented:
/test-centos-e2e-integration-main

Sunnatillo commented:
/test-centos-e2e-integration-main

furkatgofurov7 reviewed (commented):
Thanks, this is almost there to be ready, except below missing parts:

- https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/test/e2e/config/e2e_conf.yaml#L98 has to be updated to next version, `v1.5.99`
- same as previous one: https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/clusterctl-settings.json#L5
- you need to create a new metadata file for v1.5 provider in https://github.com/metal3-io/cluster-api-provider-metal3/tree/main/test/e2e/data/infrastructure-metal3

Sunnatillo commented:
/test-centos-e2e-integration-main
/test-ubuntu-e2e-integration-main

Sunnatillo commented:
/test-centos-e2e-integration-main
/test-ubuntu-e2e-integration-main

Sunnatillo commented:
/test-ubuntu-integration-main
/test-centos-integration-main
/test-centos-e2e-feature-main
/test-e2e-upgrade-main-from-release-1-4
/test-e2e-1-26-1-27-upgrade-main

Sunnatillo commented:
> Thanks, this is almost there to be ready, except below missing parts:
> 
> * https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/test/e2e/config/e2e_conf.yaml#L98 has to be updated to next version, `v1.5.99`
> * same as previous one: https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/clusterctl-settings.json#L5
> * you need to create a new metadata file for v1.5 provider in https://github.com/metal3-io/cluster-api-provider-metal3/tree/main/test/e2e/data/infrastructure-metal3

Thank you for explaining. Now I got the way how we do it.

Sunnatillo commented:
/test-ubuntu-integration-main
/test-centos-integration-main
/test-centos-e2e-feature-main
/test-e2e-upgrade-main-from-release-1-4
/test-e2e-1-26-1-27-upgrade-main

Sunnatillo commented:
/test-centos-e2e-feature-main
//test-ubuntu-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

Sunnatillo commented:
/test-centos-e2e-feature-main

kashifest commented:
/approve

adilGhaffarDev commented:
/lgtm

furkatgofurov7 commented on scripts/ci-e2e.sh:
```diff
@@ -8,14 +8,29 @@ export CAPM3PATH="${REPO_ROOT}"
 sudo chown "${USER}":"${USER}" ${WORKING_DIR}
-M3_DEV_ENV_REPO="https://github.com/metal3-io/metal3-dev-env.git"
-M3_DEV_ENV_BRANCH=main
+M3_DEV_ENV_REPO="https://github.com/Nordix/metal3-dev-env.git"
+M3_DEV_ENV_BRANCH=bump-capi-v1.5
```
We forgot to revert this, and we need to fix it

cc @Sunnatillo
